vms/platformvm/txs/mempool/mempool.go

```diff
@@ -5,42 +5,411 @@ package mempool
+		if err := m.tryEvictTx(next, gasPrice); err != nil {
+			return err
+		}
+	}
+
+	heapTx := Tx{
+		Tx:         tx,
+		Complexity: complexity,
+		gasUsed:    gasUsed,
+		GasPrice:   gasPrice,
+	}
+
+	if m.consumedUTXOs.HasOverlap(tx.InputIDs()) {
+		return fmt.Errorf("failed to add tx: %w", mempool.ErrConflictsWithOtherTx)
+	}
+
+	m.maxHeap.Push(tx.TxID, heapTx)
+	m.minHeap.Push(tx.TxID, heapTx)
+	m.consumedUTXOs.Put(tx.TxID, tx.InputIDs())
+	m.droppedTxIDs.Evict(tx.TxID)
+	m.currentGas += gasUsed
+
+	m.updateMetrics()
+
+	return nil
+}
+
+// Evict a tx if it is lower paying and the resulting total gas is within
+// capacity
+func (m *Mempool) tryEvictTx(next gas.Gas, gasPrice float64) error {
+	if _, low, ok := m.minHeap.Peek(); ok &&
+		low.GasPrice < gasPrice &&
+		next-low.gasUsed <= m.gasCapacity {
+		evictedTxID, _, _ := m.minHeap.Pop()
+
+		m.remove(evictedTxID)
+
+		return nil
+	}
+
+	return ErrGasCapacityExceeded
+}
+
+func (m *Mempool) updateMetrics() {
+	m.numTxsMetric.Set(float64(m.maxHeap.Len()))
+	m.gasAvailableMetric.Set(float64(m.gasCapacity - m.currentGas))
+}
+
+func (m *Mempool) Get(txID ids.ID) (Tx, bool) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return m.maxHeap.Get(txID)
+}
+
+func (m *Mempool) Remove(txID ids.ID) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.remove(txID)
+}
+
+func (m *Mempool) remove(txID ids.ID) {
+	heapTx, ok := m.maxHeap.Remove(txID)
+	if !ok {
+		return
 	}
 
-	return m.Mempool.Add(tx)
+	m.minHeap.Remove(txID)
+	m.consumedUTXOs.DeleteKey(txID)
+
+	m.currentGas -= heapTx.gasUsed
+
+	m.updateMetrics()
+}
+
+func (m *Mempool) Peek() (Tx, bool) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	_, tx, ok := m.maxHeap.Peek()
+
+	return tx, ok
+}
+
+func (m *Mempool) Iterate(f func(tx Tx) bool) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	for _, v := range m.maxHeap.Elements() {
+		if !f(v) {
+			return
+		}
+	}
+}
+
+func (m *Mempool) MarkDropped(txID ids.ID, reason error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if _, ok := m.maxHeap.Get(txID); ok {
+		return
+	}
+
+	m.droppedTxIDs.Put(txID, reason)
+}
+
+func (m *Mempool) GetDropReason(txID ids.ID) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	err, _ := m.droppedTxIDs.Get(txID)
+	return err
+}
+
+func (m *Mempool) Len() int {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return m.maxHeap.Len()
+}
+
+func (m *Mempool) meter(tx txs.UnsignedTx) (gas.Dimensions, gas.Gas, error) {
+	c, err := fee.TxComplexity(tx)
+	if err != nil {
+		return gas.Dimensions{}, 0, err
+	}
+
+	g, err := c.ToGas(m.weights)
+	if err != nil {
+		return gas.Dimensions{}, 0, err
+	}
+
+	return c, g, nil
+}
+
+func getUTXOs(avaxAssetID ids.ID, tx txs.UnsignedTx) (uint64, uint64, error) {
+	u := utxoGetter{AVAXAssetID: avaxAssetID}
+
+	if err := tx.Visit(&u); err != nil {
+		return 0, 0, err
+	}
+
+	return u.Consumed, u.Produced, nil
+}
+
+type utxoGetter struct {
+	AVAXAssetID ids.ID
+	Consumed    uint64
+	Produced    uint64
+}
```

Comment:
I think that we should restructure our fee calculations as needed to allow us to share the code between the mempool and the verification code.

I don't think this utxoGetter currently correctly calculates the fee for:
- AddPermissionlessValidatorTx
- AddPermissionlessDelegatorTx
- ConvertSubnetToL1Tx
- RegisterL1ValidatorTx
- IncreaseL1ValidatorBalanceTx